lib/propolis/src/accessors.rs

```diff
@@ -525,11 +518,33 @@ impl<T> Accessor<T> {
 
     /// Print the hierarchy that this node is a member of
     pub fn print(&self, highlight_self: bool) {
-        self.0.lock_tree(|tree, ent| {
-            tree.print(print_basic(highlight_self.then_some(ent.id)));
+        self.0.lock_tree(|tree| {
+            tree.print(print_basic(
+                highlight_self.then_some(NodeKey::from(&self.0)),
+            ));
         });
     }
 }
+impl<T> Drop for Accessor<T> {
+    /// Perform necessary `Node` clean-up in the containing tree during drop of
+    /// the Accessor.
+    ///
+    /// On first glance it would seem like this logic belongs in the [Drop] impl
+    /// for `Node`, rather than the [Accessor].  Doing it that way mostly works,
+    /// but poses a challenge: When the `Tree` is performing node adoption or
+    /// orphaning, it must reach back out into the Nodes it owns via
+    /// `TreeNode::node_ref`.  This poses a race for when the last
+    /// `Arc<Node<T>>` is dropped, either by the `Accessor`, or the logic in the
+    /// `Tree`.  When the latter wins, it still holds the tree lock, posing a
+    /// deadlock situation which is otherwise impossible to prevent.
+    ///
+    /// As such it is expected that the Accessor will perform the tree
+    /// deregistration of the node when it is being dropped.  No other
+    /// structures should hold the `Arc<Node<T>>`.
```

Comment:
Thanks for the comment, this is very helpful!

Reply:
I was (and remain) a bit uncomfortable moving that load-bearing `drop` logic into the `Accessor`, so I wanted to make the reasoning clear to future readers (including myself)